src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -129,33 +267,437 @@ public ResultSet getSchemas() throws SQLException {
+     *
+     * @param f The Supplier function to execute. This function may throw a RuntimeException that
+     *     wraps a SQLException.
+     * @param <T> The return type of the Supplier.
+     * @return The return value of the Supplier.
+     * @throws SQLException If a RuntimeException that wraps a SQLException is caught.
+     */
+    private static <T> T liftSQLException(Supplier<T> f) throws SQLException {
+        try {
+            return f.get();
+        } catch (RuntimeException e) {
+            if (e.getCause() instanceof SQLException) {
+                throw (SQLException) e.getCause();
+            }
+            throw e;
+        }
+    }
+
+    // Helper class for representing all info needed to serialize column data for the
+    // getColumns and getColumnPrivileges methods. Intended for use with toGetColumnsDoc
+    // and toGetColumnPrivilegesDoc helpers.
+    private static class GetColumnsDocInfo {
+        String dbName;
+        String collName;
+        String columnName;
+        MongoJsonSchema columnSchema;
+        boolean isRequired;
+        int idx;
+
+        GetColumnsDocInfo(
+                String dbName,
+                String collName,
+                String columnName,
+                MongoJsonSchema columnSchema,
+                boolean isRequired,
+                int idx) {
+            this.dbName = dbName;
+            this.collName = collName;
+            this.columnName = columnName;
+            this.columnSchema = columnSchema;
+            this.isRequired = isRequired;
+            this.idx = idx;
+        }
+    }
+
+    // Helper for creating BSON documents for the getColumns method. Intended for use
+    // with the getColumnsFromDB helper method which is shared between getColumns and
+    // getColumnPrivileges.
+    private BsonDocument toGetColumnsDoc(GetColumnsDocInfo i) {
+        MongoSQLColumnInfo info;
+        try {
+            info = new MongoSQLColumnInfo(i.collName, i.columnName, i.columnSchema, i.isRequired);
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+
+        int nullability = info.getNullability();
+        String isNullable =
+                nullability == ResultSetMetaData.columnNoNulls
+                        ? "NO"
+                        : nullability == ResultSetMetaData.columnNullable ? "YES" : "";
+
+        return createBottomBson(
+                new BsonElement(TABLE_CAT, new BsonString(i.dbName)),
+                new BsonElement(TABLE_SCHEM, new BsonString("")),
+                new BsonElement(TABLE_NAME, new BsonString(i.collName)),
+                new BsonElement(COLUMN_NAME, new BsonString(i.columnName)),
+                new BsonElement(DATA_TYPE, new BsonInt32(info.getJDBCType())),
+                new BsonElement(TYPE_NAME, new BsonString(info.getTableAlias())),
+                new BsonElement(COLUMN_SIZE, new BsonNull()),
```

Comment:
Not sure what to do here. The MySQL variant always returns `null` for this. (Well, it really returns `INFORMATION_SCHEMA.COLUMNS.CHARACTER_MAXIMUM_LENGTH`, but we always set that to `null` in the BIC/ADL catalog builder code.)

Reply:
```
The COLUMN_SIZE column specifies the column size for the given column. For numeric data, this is the maximum precision. For character data, this is the length in characters. For datetime datatypes, this is the length in characters of the String representation (assuming the maximum allowed precision of the fractional seconds component). For binary data, this is the length in bytes. For the ROWID datatype, this is the length in bytes. Null is returned for data types where the column size is not applicable.
```
We messed up in MySQL (oh well). I think we should return these based on types. I'm not sure what to do for string columns. Either null or figure out what max strings size in mongodb actually is (I suspect it's 4GB).